contrib/babelfishpg_tsql/src/pltsql_coerce.c

```diff
@@ -1018,99 +1003,16 @@ get_immediate_base_type_of_UDT(PG_FUNCTION_ARGS)
-
-	for (int special_func_idx = 0; special_func_idx < TOTAL_TSQL_SPECIAL_FUNCTION_COUNT; special_func_idx++)
-	{
-		for (int arg_idx = 0; arg_idx < tsql_special_function_list[special_func_idx].nargs; arg_idx++)
-		{
-			for (int valid_type_idx = 0; valid_type_idx < tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].len; valid_type_idx++)
-			{
-				if (!OidIsValid(tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].valid_types_oid[valid_type_idx]))
-				{
-					type_id = (*common_utility_plugin_ptr->get_tsql_datatype_oid)(tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].valid_types[valid_type_idx]);
-
-					if (OidIsValid(type_id))
-					{
-						tsql_special_function_list[special_func_idx].valid_arg_types[arg_idx].valid_types_oid[valid_type_idx] = type_id;
-					}
-					else
-					{
-						/* type id is not loaded. wait for next scan */
-						inited_tsql_special_function_list = false;
-					}
-				}
-			}
-		}
-	}
-}
-
-/*
- * For variadic function, validate the input argument data types.
- */
-static bool
-validate_variadic_special_function_args(char *func_name, List* fargs, int nargs, int nfixed_args, Oid *input_typeids)
-{
-	Oid		input_type_id, base_type_id;
-	Oid		sys_varcharoid = get_sys_varcharoid();
-
-	if ((strlen(func_name) == 6 && strncmp(func_name, "concat", 6) == 0) ||
-		 (strlen(func_name) == 9 && strncmp(func_name, "concat_ws", 9) == 0))
-	{
-		for (int i = nfixed_args; i < nargs; i++)
-		{
-			if (input_typeids[i] == UNKNOWNOID)
-			{
-				Node *arg = (Node *) lfirst(list_nth_cell(fargs, i));
-				if (IsA(arg, Const) && ((Const *)arg)->constisnull)
-					continue;
-				else
-					input_type_id = sys_varcharoid;
-			}
-			else
-				input_type_id = input_typeids[i];
-
-			base_type_id = get_immediate_base_type_of_UDT_internal(input_type_id);
-			if (OidIsValid(base_type_id))
-				input_type_id = base_type_id;
-
-			/* Check if the input type is implicitly castable to varchar. */
-			if (!can_coerce_type(1, &input_type_id, &sys_varcharoid, COERCION_IMPLICIT))
-				ereport(ERROR,
-					(errcode(ERRCODE_UNDEFINED_FUNCTION),
-						errmsg("Implicit conversion from data type %s to varchar is not allowed.", format_type_be(input_typeids[i]))));
-		}
-		return true;
-	}
-	return false;
-}
-
 /*
  * For a given function details, validate whether it is in special function list
- * and also validate the input argument data types.
  */
 bool
-validate_special_function(char *func_nsname, char *func_name, List* fargs, int nargs, Oid *input_typeids, bool num_args_match)
+validate_special_function(char *func_nsname, char *func_name, int nargs, bool num_args_match)
 {
 	tsql_special_function_t    *special_func;
-	bool                        type_match;
-	Oid                         input_type_id, valid_type_id, base_type_id;
-	Oid                         sys_varcharoid;
 
 	/* Sanity checks */
-	if (func_name == NULL || (nargs != 0 && input_typeids == NULL) || fargs == NIL)
```

Comment:
previously we had a check `(nargs != 0 && input_typeids == NULL)` to make sure that `input_typeids` is not null when `nargs` is 0, now since we are not using `input_typeids` in `validate_special_function`, we do not need this check.